Implement JavaInstanceSpecLoader for exonum-launcher [ECR-3807]

vitvakatu opened the pull request:
## Overview

Implemented JavaInstanceSpecLoader plugin for exonum-launcher. It allows user to specify path to protobuf sources of the configuration message and therefore allows to:
1. Instantiate service for Java runtime which does not support ProtobufSources
2. Instantiate any other service of any other runtime that uses protobuf for configuration messages serialization (Rust runtime included).

The plugin is covered with tests, but was not tested in real world as exonum-launcher is a little outdated right now. You're welcome with additional test cases to implement.

Questions:
1. Why `data` field is needed? 
  Due to current implementation of exonum-launcher, the best way to represent an arbitrary config struct and additional parameters for the plugin is additional `data` field. The name is arguable though.
2. Why is it called `JavaInstanceSpecLoader`?
  Indeed, there isn't much Java in this plugin. It might be actually be moved to exonum-launcher as default or semi-default plugin later, but it is initially developed by EJB team and for EJB purposes.


TODO: 
1. Implement tests for JavaDeploySpecLoader as well (in a different task) (ECR-3870)
2. Set up travis for plugins directory (ECR-3870)
3. ~~Do smth with accidental commit of changes in `proto` submodule. (how?)~~
4. ~~Rename directory `launcher-plugin` -> `launcher-plugins`~~
---
See: https://jira.bf.local/browse/ECR-3807

### Definition of Done

- [ ] There are no TODOs left in the code
- [ ] Change is covered by automated [tests](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#tests)
- [ ] The [coding guidelines](https://github.com/exonum/exonum-java-binding/blob/master/CONTRIBUTING.md#the-code-style) are followed
- [ ] Public API has Javadoc
- [ ] Method preconditions are checked and documented in the Javadoc of the method
- [ ] Changelog is updated if needed (in case of notable or breaking changes)
- [ ] The continuous integration build passes

popzxc commented on exonum-java-binding/launcher-plugins/exonum_java_instance_plugin/plugin.py:
```diff
@@ -0,0 +1,94 @@
+from exonum_client.protobuf_loader import ProtobufLoader
+from exonum_client.protoc import Protoc
+from exonum_launcher.instances import InstanceSpecLoader, InstanceSpecLoadError
+from importlib import import_module
```
Nit: imports from stdlib should be placed before external imports.

popzxc commented on exonum-java-binding/launcher-plugins/exonum_java_instance_plugin/plugin.py:
```diff
@@ -0,0 +1,94 @@
+        result = config_encoder(instance.config[_DATA_FIELD_NAME])
+
+        # Remove the generated temporary directory:
+        self.cleanup(tmp_dir)
```
Nit: if exception will be raised above, cleanup won't be performed. You can use [`contextlib.contextmanager`](https://docs.python.org/3/library/contextlib.html#contextlib.contextmanager) decorator to prevent this without creating an additional class.

popzxc commented on exonum-java-binding/launcher-plugins/exonum_java_instance_plugin/plugin.py:
```diff
@@ -0,0 +1,94 @@
+        return result
+
+    @staticmethod
+    def cleanup(tmp_dir):
```
Nit here and below: it's a good practice to add type hints, even if function returns `None`, so `mypy` will be able to analyze code better (explicit `None` return type means that we know that function **should** return nothing and not just forgot to add return type). This is completely optional though.

popzxc commented on exonum-java-binding/launcher-plugins/tests/test_data/proto_sources/service.proto:
```diff
@@ -0,0 +1,21 @@
+package exonum.examples.timestamping;
+
+message Config {
+    string time_service_name = 2;
```
Why the number is `2`?

vitvakatu replied:
To match the example binary representation from Protobuf docs. I'll add an explanation here.

dmitry-timofeev replied:
The binary representation is unrelated to field numbers (but each field does include the field number, on top of the field _type_). The field number has nothing to do with the on-the wire encoding (the _type_ = string defines the wire encoding, the tag — identifies the field): https://developers.google.com/protocol-buffers/docs/proto3#assigning-field-numbers

dmitry-timofeev commented:
> exonum-launcher is a little outdated right now. 

What do you mean by that? Is it no longer compatible with the recent Exonum revisions after any changes?

dmitry-timofeev commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,94 @@
+plugins:
+  runtime: {}
+  artifact: 
+    service_instance_name: "exonum_java_instance_plugin.JavaInstanceSpecLoader"
```
service_instance_name? What do service instances have to do with the plugin?

vitvakatu replied:
My mistake, it must be `service_artifact_name`, to bind a specific plugin to a specific artifact.

dmitry-timofeev commented on exonum-java-binding/launcher-plugins/README.md:
````diff
@@ -0,0 +1,94 @@
+    service_instance_name: "exonum_java_instance_plugin.JavaInstanceSpecLoader"
+```
+
+To instantiate a service with a custom configuration you need to create Protobuf
````
create — I think, _take_ a service configuration message protobuf definition (as it is defined by the service and must be used as is).

dmitry-timofeev commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,94 @@
+
+To instantiate a service with a custom configuration you need to create Protobuf
+source of the configuration message and place it in specific directory. The name 
+of the message must be `Config`:
```
Is it a requirement? Or convention-over-configuration thing? The QA service currently uses the InitialConfiguration name.

We currently don't support re-configuration, but that interface is allowed to have a different message for configuration updates (though I don't think it is very likely). Will the need to support two different 'configuration' messages affect this restriction?

vitvakatu replied:
It's possible to relax such restriction, thanks to Python

dmitry-timofeev commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,94 @@
+
+- `sources`. Points to a directory with Protobuf-sources of service configuration 
+message. We use `proto_sources` directory.
+- `module_name`. A name (without extension) of the file where `Config` message 
```
Would a full filename as 'config_message_source' or 'config_filename' be easier to understand than a 'module_name'?

vitvakatu replied:
Probably yes, naming suggestions are always welcomed

dmitry-timofeev commented on exonum-java-binding/launcher-plugins/README.md:
````diff
@@ -0,0 +1,94 @@
+
+```yaml
+instances:
+  service_instance_name:
````
Why does 'service_instance_name' have anything but a string service instance name?

dmitry-timofeev replied:
Or is the key = instance name?

vitvakatu replied:
Yep, the key is the instance name, and the value is additional parameters

popzxc commented:
> What do you mean by that? Is it no longer compatible with the recent Exonum revisions after any changes?

Exonum has got several breaking changes (e.g. moving proto endpoint into the rust runtime) and will get one more soon (https://github.com/exonum/exonum/pull/1538), so the Exonum python client is a bit outdated now. It will be fixed in a couple of days I hope.

vitvakatu commented:
Addressed review comments, added bunch of improvements:
1. `module_name` -> `config_message_source`
2. Fixed bug with broken multiple service instances in tests (thanks to @popzxc )
3. Added `message_name` optional parameter

dmitry-timofeev commented on exonum-java-binding/launcher-plugins/tests/test_data/custom_message_name.yml:
```diff
@@ -0,0 +1,26 @@
+
+plugins:
+  artifact:
+    cryptocurrency: "exonum_java_instance_plugin.JavaInstanceSpecLoader"
```
Where is the 'cryptocurrency' artifact used? I only see timestamping. Isn't it redundant?

dmitry-timofeev commented on exonum-java-binding/launcher-plugins/tests/test_data/proto_sources/custom_message_name.proto:
```diff
@@ -0,0 +1,22 @@
+package exonum.examples.timestamping;
+
+message InitialConfig {
+    // Corresponds to https://developers.google.com/protocol-buffers/docs/encoding#strings
```
What does correspond to string encoding?

2 is a field tag and has nothing to do with the on-the wire encoding (the type defines the wire encoding, the tag — identifies the field): https://developers.google.com/protocol-buffers/docs/proto3#assigning-field-numbers

vitvakatu replied:
I really can't understand why the `2` became such an important detail. It is used just to use the same binary representation as in the example under the provided link.

Changed to `1` and removed links to protobuf docs.

dmitry-timofeev commented on exonum-java-binding/launcher-plugins/tests/test_instance_plugin.py:
```diff
@@ -0,0 +1,46 @@
+        instance_loader = JavaInstanceSpecLoader()
+
+        for instance in config.instances:
+            spec = instance_loader.load_spec(None, instance)
```
What is `spec`? There are a handful of various "specs" in the deploy & start process. Is it an initial configuration?

The core we use uses `ArtifactSpec.payload` field name, which could be improved, I think (initial_config?): exonum-java-binding/common/src/main/proto/src/runtime.proto:45

vitvakatu replied:
🤔 Isn't the `ArtifactSpec.payload` contain binary representation of the _artifact_ (in our case, a string containing filename of the artifact)?

InstanceLoader.load_spec returns a serialized initial service configuration.

dmitry-timofeev commented on exonum-java-binding/launcher-plugins/tests/test_instance_plugin.py:
```diff
@@ -0,0 +1,46 @@
+        for instance in config.instances:
+            spec = instance_loader.load_spec(None, instance)
+            # https://developers.google.com/protocol-buffers/docs/encoding#strings
+            self.assertEqual(spec, b"\x12\x07\x74\x65\x73\x74\x69\x6e\x67")
```
Given the loop uses the same constant and a single instance is expected, processing that single instance would be more straightforward and won't give a false impression you expect to have several instances with same configurations (timestamping_instance/instance = config.instances[0]).

skletsun commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,96 @@
+
+## Usage
+
+TODO: Move the entire section to the EJB App tutorial.
```
Is this going to be implemented within this PR?

vitvakatu replied:
Probably not, the entire readme isn't even a part of this task.

vitvakatu commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -53,19 +53,18 @@ runtimes:
   runtime: {}
   artifact: 
-    service_instance_name: "exonum_java_instance_plugin.JavaInstanceSpecLoader"
+    service_artifact_name: "exonum_java_instance_plugin.JavaInstanceSpecLoader"
```
The key is the artifact identifier from `artifacts` section of the config.
```yaml
artifacts:
  service_artifact_name:
    runtime: java
    name: "com.exonum.examples:service-name:0.9.0-SNAPSHOT"
    spec:
      artifact_filename: "service-artifact-filename.jar"
```

Added a sample config and will probably try to rephrase readme cause it's not obvious for reader.

dmitry-timofeev replied:
Ok, so this format requires a second name for the artifact (in the example config, same as instance names). Is it worth it to add a second, shorter name instead of using the name an artifact already has?

vitvakatu replied:
ping @popzxc 
The plugin does not change this part of the exonum-launcher. Though such implementation should be possible, as YAML supports strings as keys.

popzxc replied:
Plugins section has the following format:
```yaml
plugins:
  runtime:
     $RUNTIME_NAME: $RUNTIME_PLUGIN_PYTHON_CLASS
  artifact:
    $SPEC_ARTIFACT_ALIAS: $SPEC_PLUGIN_PYTHON_CLASS
```

So, for example let's take this config:
```yaml
artifacts:
  cryptocurrency:
    runtime: sample_runtime
    name: "exonum-cryptocurrency:0.12.0"
  
instances:
  xnm-token:
    artifact: cryptocurrency
```

Here, we want to deploy artifact `cryptocurrency`. `cryptocurrency` here is a local artifact definition to be used within config (actual spec is stored inside, see `"exonum-cryptocurrency:0.12.0"`).

If we want to provide a plugin for a runtime, it will be:

```yaml
plugins:
  runtime:
     sample_runtime: our.package.SampleRuntimeSpecLoader
```

if we want also the spec for `cryptocurrency` artifact be loaded with plugin, then plugin section will become the following:

```yaml
plugins:
  runtime:
     sample_runtime: our.package.SampleRuntimeSpecLoader
  artifact:
    cryptocurrency: our.package.CryptocurrencySpecLoader
```

Why is it done that way? At least for `rust` runtime artifact names contain a version. So in order to just bump a version, you will have to update it everywhere (so it's up to N + 1 times -- N for every instance of that artifact in config, and 1 for plugin), with aliases it will be only one place.

And IMO it's just more readable.

dmitry-timofeev replied:
Thanks for clarifications, it makes sense!

dmitry-timofeev commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,138 @@
+# Java Runtime Plugins for [Exonum Launcher](https://github.com/exonum/exonum-launcher)
+
+- Java Runtime Plugin allows to deploy and start Java services.
+- Java Instance Plugin allows to use an arbitrary Protobuf message for service 
```
Local Configuration Plugin? Instance Configuration Plugin? (Instance) Initial Configuration Plugin?

It is not Java-specific, as it says below.

vitvakatu replied:
Should I also rename the python package name?

dmitry-timofeev replied:
If that makes sense — please do.

vitvakatu replied:
Kinda hard to decide what name suits better. @popzxc @skletsun @dip56 what do you think?

popzxc replied:
I have no particular opinion on this topic.

dmitry-timofeev commented on exonum-java-binding/launcher-plugins/README.md:
````diff
@@ -0,0 +1,138 @@
+    private-api-port: 8081
+```
+
+You need to specify every node in the network.
````
Does it require the private ports of nodes to be accessible from a single location? I wonder if there are any other operation on the private port that do not authorize the user, making such setup impractical in distributed scenarious. Is that a known, deliberate restriction?

vitvakatu replied:
ping @popzxc

popzxc replied:
`exonum-launcher` provides functionality to interact with nodes that you own. If you own all the nodes, you can send all the requests from one launch.
If network has several administrators, every administrator has to run `exonum-launcher` with the networks he owns.

Decentralized supervisor (as the oppose to the simple supervisor, which requires 1 request only, and which is now default) requires N requests and `exonum-launcher` is *trying* to make the life of the developer better, but this is not a generic solution for every situation.

dmitry-timofeev replied:
Thanks for clarifications!

@vitvakatu , The documentation seems incomplete and has to be updated then, doesn't it?

vitvakatu replied:
@dmitry-timofeev it seems, I will update.

dmitry-timofeev commented:
@Maria-Nosyk Would you review please the instructions in readme?

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
````diff
@@ -0,0 +1,138 @@
+    name: "exonum-timestamping:0.12.0"
+```
+
+Add `plugins` section to enable both Java Runtime plugin and Java Instance plugin. Runtime plugin is
````
```suggestion
Add a `plugins` section to enable both Java Runtime plugin and Java Instance plugin. Runtime plugin is
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
````diff
@@ -0,0 +1,138 @@
+```
+
+In our example we will use Java Instance plugin to serialize initial configuration parameters of
+`timestamping` service in Protobuf. We need to take a `service.proto` file with message
````
```suggestion
the `timestamping` service in Protobuf. We need to take a `service.proto` file with the message
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,138 @@
+
+In our example we will use Java Instance plugin to serialize initial configuration parameters of
+`timestamping` service in Protobuf. We need to take a `service.proto` file with message
+description from service sources and place it inside some known directory.
```
```suggestion
description from the service sources and place it inside some known directory.
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
````diff
@@ -0,0 +1,138 @@
+    timestamping: "exonum_java_instance_plugin.JavaInstanceSpecLoader"
+```
+
+In our example we will use Java Instance plugin to serialize initial configuration parameters of
````
```suggestion
In our example we will use the Java Instance plugin to serialize initial configuration parameters of
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
````diff
@@ -0,0 +1,138 @@
+  }
+  ```
+
+Finally, add `instances` section that describes the list of service instances to be started in the
````
```suggestion
Finally, add an `instances` section that describes the list of service instances you want to start in the
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
````diff
@@ -0,0 +1,138 @@
+  ```
+
+Finally, add `instances` section that describes the list of service instances to be started in the
+blockchain. For each instance you need to specify artifact name alias. Java Instance plugin also
````
```suggestion
blockchain. For each instance you need to specify its artifact name alias. Java Instance plugin also
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,138 @@
+
+Finally, add `instances` section that describes the list of service instances to be started in the
+blockchain. For each instance you need to specify artifact name alias. Java Instance plugin also
+requires a list of additional parameters, which we provide for `timestamping` instance:
```
```suggestion
requires a list of additional parameters, which we provide for the `timestamping` instance:
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,138 @@
+blockchain. For each instance you need to specify artifact name alias. Java Instance plugin also
+requires a list of additional parameters, which we provide for `timestamping` instance:
+
+- `sources`. Points to a directory with Protobuf-sources of service configuration 
```
```suggestion
- `sources`. Points to a directory with the Protobuf sources of the service configuration 
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,138 @@
+requires a list of additional parameters, which we provide for `timestamping` instance:
+
+- `sources`. Points to a directory with Protobuf-sources of service configuration 
+message. We use `proto_sources` directory.
```
```suggestion
message. We use the `proto_sources` directory.
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,138 @@
+
+- `sources`. Points to a directory with Protobuf-sources of service configuration 
+message. We use `proto_sources` directory.
+- `config_message_source`. A file name where `message_name` message 
```
```suggestion
- `config_message_source`. A file name where the `message_name` message 
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,138 @@
+- `sources`. Points to a directory with Protobuf-sources of service configuration 
+message. We use `proto_sources` directory.
+- `config_message_source`. A file name where `message_name` message 
+is located. In our example we use `service.proto` file.
```
```suggestion
is located. In our example we use the `service.proto` file.
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,138 @@
+message. We use `proto_sources` directory.
+- `config_message_source`. A file name where `message_name` message 
+is located. In our example we use `service.proto` file.
+- `message_name`. A name of the Protobuf message used to represent service configuration.
```
```suggestion
- `message_name`. A name of the Protobuf message used to represent the service configuration.
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
```diff
@@ -0,0 +1,138 @@
+is located. In our example we use `service.proto` file.
+- `message_name`. A name of the Protobuf message used to represent service configuration.
+  Optional, defaults to `Config`.
+- `data`. Your actual configuration in the format corresponding to `message_name` message.
```
```suggestion
- `data`. Your actual configuration in the format corresponding to the `message_name` message.
```

Maria-Nosyk commented on exonum-java-binding/launcher-plugins/README.md:
````diff
@@ -0,0 +1,138 @@
+        time_service_name: "time"
+```
+
+See [sample-config.yml](sample-config.yml) for the final state of configuration file.
````
```suggestion
See [sample-config.yml](sample-config.yml) for the final state of the configuration file.
```
